packages/maskbook/src/web-workers/OnDemandWorker.ts

```diff
@@ -0,0 +1,34 @@
+type ctor = [stringUrl: string | URL, options?: WorkerOptions]
+export class OnDemandWorker extends Worker {
+    protected readonly init: ctor
+    protected worker: Worker | undefined = this
+    protected readonly id = Math.random().toString(16).slice(2)
+    constructor(...init: ctor) {
+        super(...init)
+        this.init = init
+        this.log('created with', ...init)
+    }
+    protected log(...args: any[]) {
+        console.log(`OnDemandWorker ${this.id}`, ...args)
+    }
+    protected lastUsed = Date.now()
+    protected use(onReady: () => void) {
+        this.lastUsed = Date.now()
+        this.log('kept alive')
+
+        if (this.worker) return onReady()
+        this.worker = new Worker(...this.init)
+        this.worker.addEventListener('error', (e) => this.dispatchEvent(e))
```

Comment:
No need for removing those listeners?

Reply:
![](https://user-images.githubusercontent.com/5390719/99031892-1f557500-25b3-11eb-859e-d84f3c700e60.png)

After the EventTarget is being recycled, its listeners are also recycled so doesn't count to reference to the Uint8Array.